File: src/graph/Color.js
import parsegraph_TestSuite from '../TestSuite';

export default function parsegraph_Color(r, g, b, a)
{
    if(arguments.length === 1) {
        g = r;
        b = r;
        a = 1;
    }
    else if(arguments.length === 3) {
        a = 1;
    }
    else if(arguments.length !== 4) {
        throw new Error("Color must be given initial component values.");
    }
    this._r = Math.min(1, Math.max(0, r));
    this._g = Math.min(1, Math.max(0, g));
    this._b = Math.min(1, Math.max(0, b));
    this._a = Math.min(1, Math.max(0, a));
}

parsegraph_Color.prototype.r = function() {
    return this._r;
};

parsegraph_Color.prototype.g = function() {
    return this._g;
};

parsegraph_Color.prototype.b = function() {
    return this._b;
};

parsegraph_Color.prototype.a = function() {
    return this._a;
};

parsegraph_Color.prototype.setA = function(value) {
    this._a = Math.min(1, Math.max(0, value));
    return this;
};

parsegraph_Color.prototype.setR = function(value) {
    this._r = Math.min(1, Math.max(0, value));
    return this;
};

parsegraph_Color.prototype.setG = function(value) {
    this._g = Math.min(1, Math.max(0, value));
    return this;
};

parsegraph_Color.prototype.setB = function(value) {
    this._b = Math.min(1, Math.max(0, value));
    return this;
};

parsegraph_Color.prototype.multiply = function(other) {
    return new parsegraph_Color(
        this.r() * other.r(),
        this.g() * other.g(),
        this.b() * other.b(),
        this.a() * other.a()
    );
};

parsegraph_Color.prototype.premultiply = function(other) {
    return new parsegraph_Color(
        (this.a() * this.r()) + (other.r() * (1.0 - this.a())),
        (this.a() * this.g()) + (other.g() * (1.0 - this.a())),
        (this.a() * this.b()) + (other.b() * (1.0 - this.a())),
        1.0
    );
};

parsegraph_Color.prototype.asRGB = function() {
    return "rgb(" +
        Math.round(this._r * 255) + ", " +
        Math.round(this._g * 255) + ", " +
        Math.round(this._b * 255) + ")"
};

export function parsegraph_inverseSRGBCompanding(v)
{
    if(v <= 0.04045) {
        return v/12.92;
    }
    return Math.pow((v+0.055)/1.055, 2.4);
}

export function parsegraph_SRGBCompanding(v)
{
    if(v <= 0.0031308) {
        return v*12.92;
    }
    return 1.055*Math.pow(v,1/2.4)-0.055;
}

parsegraph_Color.prototype.luminance = function() {
    // sRGB color model.
    var x1 = parsegraph_inverseSRGBCompanding(this.r());
    var y1 = parsegraph_inverseSRGBCompanding(this.g());
    var z1 = parsegraph_inverseSRGBCompanding(this.b());
    return x1 * 0.648431 + y1 * 0.321152 + z1 * 0.155886;
};

parsegraph_Color.prototype.interpolate = function(other, interp) {
    //console.log("Interpolating");
    interp = Math.min(1, Math.max(0, interp));

    var e = 216/24389;
    var k = 24389/27;

    //console.log("r=" + this.r() + ", g=" + this.g()+ ", b=" + this.b());
    var x1 = parsegraph_inverseSRGBCompanding(this.r());
    var y1 = parsegraph_inverseSRGBCompanding(this.g());
    var z1 = parsegraph_inverseSRGBCompanding(this.b());
    //console.log("x1=" + x1 + ", y1=" + y1 + ", z1=" + z1);

    var xref1 = x1*0.648431;
    var yref1 = y1*0.321152;
    var zref1 = z1*0.155886;

    var fx1;
    if(xref1 > e) {
        fx1 = Math.pow(xref1, 1/3);
    }
    else {
        fx1 = (k*xref1+16)/116;
    }
    var fy1;
    if(yref1 > e) {
        fy1 = Math.pow(yref1, 1/3);
    }
    else {
        fy1 = (k*yref1+16)/116;
    }
    var fz1;
    if(zref1 > e) {
        fz1 = Math.pow(zref1, 1/3);
    }
    else {
        fz1 = (k*zref1+16)/116;
    }

    var L1 = 116*fy1-16;
    var a1 = 500*(fx1-fy1);
    var b1 = 200*(fy1-fz1);
    //console.log("L=" + L1 + ", a1=" + a1 + ", b1=" + b1);

    var C1 = Math.sqrt(Math.pow(a1, 2) + Math.pow(b1, 2));
    var H1 = Math.atan2(a1, b1);
    if(H1 < 0) {
        H1 += 2 * Math.PI;
    }

    //console.log("L=" + L1 + ", C1=" + C1 + ", H1=" + H1);

    var x2 = parsegraph_inverseSRGBCompanding(other.r());
    var y2 = parsegraph_inverseSRGBCompanding(other.g());
    var z2 = parsegraph_inverseSRGBCompanding(other.b());

    var xref2 = x2/0.648431;
    var yref2 = y2/0.321152;
    var zref2 = z2/0.155886;

    var fx2;
    if(xref2 > e) {
        fx2 = Math.pow(xref2, 1/3);
    }
    else {
        fx2 = (k*xref2+16)/116;
    }
    var fy2;
    if(yref2 > e) {
        fy2 = Math.pow(yref2, 1/3);
    }
    else {
        fy2 = (k*yref2+16)/116;
    }
    var fz2;
    if(zref2 > e) {
        fz2 = Math.pow(zref2, 1/3);
    }
    else {
        fz2 = (k*zref2+16)/116;
    }
    var L2 = 116*fy2-16;
    var a2 = 500*(fx2-fy2);
    var b2 = 200*(fy2-fz2);

    var C2 = Math.sqrt(Math.pow(a2, 2) + Math.pow(b2, 2));
    var H2 = Math.atan2(a2, b2);
    if(H2 < 0) {
        H2 += 2 * Math.PI;
    }
    //console.log("L2=" + L2 + ", C2=" + C2 + ", H2=" + H2);

    var L3 = L1 + (L2 - L1) * interp;
    var C3 = C1 + (C2 - C1) * interp;
    var H3 = H1 + (H2 - H1) * interp;
    //console.log("L3=" + L3 + ", C3=" + C3 + ", H3=" + H3);

    var a3 = C3 * Math.cos(H3);
    var b3 = C3 * Math.sin(H3);
    //console.log("L3=" + L3 + ", a3=" + a3 + ", b3=" + b3);

    var fy3 = (L3 + 16)/116;
    var fz3 = fy3 - b3/200;
    var fx3 = a3/500+fy3;

    var zref3 = Math.pow(fz3, 3);
    if(zref3 <= e) {
        zref3 = (116*fz3-16)/k;
    }

    var yref3;
    if(L3 > k * e) {
        yref3 = Math.pow((L3+16)/116, 3);
    }
    else {
        yref3 = L3/k;
    }

    var xref3 = Math.pow(fx3, 3);
    if(xref3 <= e) {
        xref3 = (116*fx3-16)/k;
    }

    var x3 = xref3*0.648431;
    var y3 = yref3*0.321152;
    var z3 = zref3*0.155886;
    //console.log("x3=" + x3 + ", y3=" + y3 + ", z3=" + z3);

    return new parsegraph_Color(
        parsegraph_SRGBCompanding(x3),
        parsegraph_SRGBCompanding(y3),
        parsegraph_SRGBCompanding(z3),
        this.a() + (other.a() - this.a()) * interp
    );
};

export function parsegraph_fromRGB(rgb, defaultAlpha)
{
    // Default alpha to 255.
    if(arguments.length === 1) {
        defaultAlpha = 255;
    }

    // Extract the color from the string, as formatted in asRGB.
    var value = [];
    rgb.trim().substring("rgb(".length, rgb.length - 1).split(',').forEach(function(c) {
        value.push(c.trim() - 0);
    });
    if(value.length < 3) {
        throw new Error("Failed to parse color");
    }
    if(value.length === 3) {
        value.push(defaultAlpha);
    }

    // Return a new color.
    return new parsegraph_Color(value[0]/255, value[1]/255, value[2]/255, value[3]/255);
};

parsegraph_Color.prototype.clone = function() {
    return parsegraph_createColor(this.r(), this.g(), this.b(), this.a());
};

parsegraph_Color.prototype.equals = function(other) {
    return this.r() === other.r() && this.g() === other.g() && this.b() === other.b() && this.a() === other.a();
};

export function parsegraph_createColor(r, g, b, a)
{
    return new parsegraph_Color(r, g, b, a);
}

const parsegraph_Color_Tests = new parsegraph_TestSuite("parsegraph_Color");

parsegraph_Color_Tests.addTest("parsegraph_Color.simplify", function() {
});

/*parsegraph_Color_Tests.addTest("parsegraph_Color.interpolate trivial", function() {
    var r = new parsegraph_Color(0, 0, 1);
    var b = new parsegraph_Color(1, 1, 0);
    var c = r.interpolate(b, 0);
    if(!c.equals(r)) {
        return "Trivial interpolate (interp=0) does not work: " + c.asRGB();
    }
});

parsegraph_Color_Tests.addTest("parsegraph_Color.interpolate trivial", function() {
    var r = new parsegraph_Color(0, 0, 1);
    var b = new parsegraph_Color(1, 1, 0);
    var c = r.interpolate(b, 1);
    if(!c.equals(b)) {
        return "Trivial interpolate (interp=1) does not work: " + c.asRGB();
    }
});

parsegraph_Color_Tests.addTest("parsegraph_Color.interpolate", function() {
    var r = new parsegraph_Color(0, 0, 1);
    var b = new parsegraph_Color(1, 1, 0);
    var c = r.interpolate(b, 0);
    if(!c.equals(new parsegraph_Color(0, 1, 0))) {
        return "Colors do not interpolate properly: " + c.asRGB();
    }
});
*/

File: src/graph/Color.c
#include "Color.h"
#include <stdlib.h>
#include <stdio.h>
#include <math.h>
#include "parsegraph_math.h"

// Background
float parsegraph_BACKGROUND_COLOR[] = {
    0, 47.0/255, 57.0/255, 1
   // 255/255, 255/255, 255/255, 1
    //45.0/255, 84.0/255, 127.0/255, 1
};

float* parsegraph_Color_new(apr_pool_t* pool, float r, float g, float b, float a)
{
    float* color;
    if(pool) {
        color = apr_palloc(pool, sizeof(float)*4);
    }
    else {
        color = malloc(sizeof(float)*4);
    }
    parsegraph_Color_SetRGBA(color, r, g, b, a);
    return color;
}

void parsegraph_Color_destroy(float* color)
{
    free(color);
}

void parsegraph_Color_copy(float* dest, float* src)
{
    memcpy(dest, src, sizeof(float)*4);
}

void parsegraph_Color_SetRGBA(float* color, float r, float g, float b, float a)
{
    color[0] = r;
    color[1] = g;
    color[2] = b;
    color[3] = a;
}

void parsegraph_Color_setRGBA(float* color, float r, float g, float b, float a)
{
    parsegraph_Color_SetRGBA(color, r, g, b, a);
}

void parsegraph_Color_SetRGB(float* color, float r, float g, float b)
{
    color[0] = r;
    color[1] = g;
    color[2] = b;
}

void parsegraph_Color_setRGB(float* color, float r, float g, float b)
{
    parsegraph_Color_SetRGB(color, r, g, b);
}

void parsegraph_Color_multiply(float* sink, float* src, float* other)
{
    sink[0] = src[0] * other[0];
    sink[1] = src[1] * other[1];
    sink[2] = src[2] * other[2];
    sink[3] = src[3] * other[3];
}

void parsegraph_Color_premultiply(float* sink, float* src, float* other)
{
    sink[0] = (src[3] * src[0]) + (other[0] * (1.0 - src[3]));
    sink[1] = (src[3] * src[1]) + (other[1] * (1.0 - src[3]));
    sink[2] = (src[3] * src[2]) + (other[2] * (1.0 - src[3]));
    sink[3] = 1.0;
};

int parsegraph_Color_asRGB(float* c, char* buf, size_t len)
{
    return snprintf(buf, len, "rgb(%0.2f, %0.2f, %0.2f)",
        roundf(c[0]*255),
        roundf(c[1]*255),
        roundf(c[2]*255)
    );
}

float parsegraph_inverseSRGBCompanding(float v)
{
    if(v <= 0.04045) {
        return v/12.92;
    }
    return powf((v+0.055)/1.055, 2.4);
}

float parsegraph_SRGBCompanding(float v)
{
    if(v <= 0.0031308) {
        return v*12.92;
    }
    return 1.055*powf(v,1/2.4)-0.055;
}

float parsegraph_Color_luminance(float* c)
{
    // sRGB color model.
    float x1 = parsegraph_inverseSRGBCompanding(c[0]);
    float y1 = parsegraph_inverseSRGBCompanding(c[1]);
    float z1 = parsegraph_inverseSRGBCompanding(c[2]);
    return x1 * 0.648431 + y1 * 0.321152 + z1 * 0.155886;
}

void parsegraph_Color_interpolate(float* sink, float* c, float* other, float interp)
{
    //console.log("Interpolating");
    interp = parsegraph_clamp(0, interp, 1);

    float e = 216/24389;
    float k = 24389/27;

    //console.log("r=" + this.r() + ", g=" + this.g()+ ", b=" + this.b());
    float x1 = parsegraph_inverseSRGBCompanding(c[0]);
    float y1 = parsegraph_inverseSRGBCompanding(c[1]);
    float z1 = parsegraph_inverseSRGBCompanding(c[2]);
    //console.log("x1=" + x1 + ", y1=" + y1 + ", z1=" + z1);

    float xref1 = x1*0.648431;
    float yref1 = y1*0.321152;
    float zref1 = z1*0.155886;

    float fx1;
    if(xref1 > e) {
        fx1 = powf(xref1, 1/3);
    }
    else {
        fx1 = (k*xref1+16)/116;
    }
    float fy1;
    if(yref1 > e) {
        fy1 = powf(yref1, 1/3);
    }
    else {
        fy1 = (k*yref1+16)/116;
    }
    float fz1;
    if(zref1 > e) {
        fz1 = powf(zref1, 1/3);
    }
    else {
        fz1 = (k*zref1+16)/116;
    }

    float L1 = 116*fy1-16;
    float a1 = 500*(fx1-fy1);
    float b1 = 200*(fy1-fz1);
    //console.log("L=" + L1 + ", a1=" + a1 + ", b1=" + b1);

    float C1 = sqrtf(powf(a1, 2) + powf(b1, 2));
    float H1 = atan2f(a1, b1);
    if(H1 < 0) {
        H1 += 2 * 3.14159;
    }

    //console.log("L=" + L1 + ", C1=" + C1 + ", H1=" + H1);

    float x2 = parsegraph_inverseSRGBCompanding(other[0]);
    float y2 = parsegraph_inverseSRGBCompanding(other[1]);
    float z2 = parsegraph_inverseSRGBCompanding(other[2]);

    float xref2 = x2/0.648431;
    float yref2 = y2/0.321152;
    float zref2 = z2/0.155886;

    float fx2;
    if(xref2 > e) {
        fx2 = powf(xref2, 1/3);
    }
    else {
        fx2 = (k*xref2+16)/116;
    }
    float fy2;
    if(yref2 > e) {
        fy2 = powf(yref2, 1/3);
    }
    else {
        fy2 = (k*yref2+16)/116;
    }
    float fz2;
    if(zref2 > e) {
        fz2 = powf(zref2, 1/3);
    }
    else {
        fz2 = (k*zref2+16)/116;
    }
    float L2 = 116*fy2-16;
    float a2 = 500*(fx2-fy2);
    float b2 = 200*(fy2-fz2);

    float C2 = sqrtf(powf(a2, 2) + powf(b2, 2));
    float H2 = atan2f(a2, b2);
    if(H2 < 0) {
        H2 += 2 * 3.14159;
    }
    //console.log("L2=" + L2 + ", C2=" + C2 + ", H2=" + H2);

    float L3 = L1 + (L2 - L1) * interp;
    float C3 = C1 + (C2 - C1) * interp;
    float H3 = H1 + (H2 - H1) * interp;
    //console.log("L3=" + L3 + ", C3=" + C3 + ", H3=" + H3);

    float a3 = C3 * cosf(H3);
    float b3 = C3 * sinf(H3);
    //console.log("L3=" + L3 + ", a3=" + a3 + ", b3=" + b3);

    float fy3 = (L3 + 16)/116;
    float fz3 = fy3 - b3/200;
    float fx3 = a3/500+fy3;

    float zref3 = powf(fz3, 3);
    if(zref3 <= e) {
        zref3 = (116*fz3-16)/k;
    }

    float yref3;
    if(L3 > k * e) {
        yref3 = powf((L3+16)/116, 3);
    }
    else {
        yref3 = L3/k;
    }

    float xref3 = powf(fx3, 3);
    if(xref3 <= e) {
        xref3 = (116*fx3-16)/k;
    }

    float x3 = xref3*0.648431;
    float y3 = yref3*0.321152;
    float z3 = zref3*0.155886;
    //console.log("x3=" + x3 + ", y3=" + y3 + ", z3=" + z3);

    sink[0] = parsegraph_SRGBCompanding(x3);
    sink[1] = parsegraph_SRGBCompanding(y3);
    sink[2] = parsegraph_SRGBCompanding(z3);
    sink[3] = c[3] + (other[3] - c[3]) * interp;
}

int parsegraph_Color_equals(float* a, float* b)
{
    return a[0] == b[0] && a[1] == b[1] && a[2] == b[2] && a[3] == b[3];
};
